waveorder/util.py

```diff
@@ -871,52 +872,180 @@ def Dual_variable_ADMM_TV_deconv_2D(AHA, b_vec, rho, lambda_u, lambda_p, itr, ve
+        reg_x[1] = calc_golden_x(reg_x[0], reg_x[3])
+        reg_x[2] = reg_x[0] + (reg_x[3]-reg_x[1])
+
+        # holds the 4 current points
+        curr_pts = []
+        for i in range(4):
+            curr_pts.append(eval_L_curve(reg_x[i]))
+        
+        opt_list = [] # save optimal point from each iteration
+        itr = 0
+        search_range = curr_pts[3].reg_x - curr_pts[0].reg_x
+        while search_range > epsilon_auto:
+            C1 = menger_curvature(curr_pts[:3])
+            C2 = menger_curvature(curr_pts[1:])
+            
+            # make sure right curvature is positive!
+                # keep moving the right edge inwards (and adjusting the middle points accordingly)
+            while C2 < 0:
+                new_reg_x = calc_golden_x(curr_pts[0].reg_x, curr_pts[2].reg_x)
+                curr_pts[3] = curr_pts[2]
+                curr_pts[2] = curr_pts[1]
+                curr_pts[1] = eval_L_curve(new_reg_x)
+                C2 = menger_curvature(curr_pts[1:])
+            C1 = menger_curvature(curr_pts[:3])
+
+            # case 1: left 3 points are better
+                # [a, b, c, d] --> [a, (a*phi+c)/(1+phi), b, c]
+            if C1 > C2:
+                opt_list.append(curr_pts[1])
+                new_reg_x = calc_golden_x(curr_pts[0].reg_x, curr_pts[2].reg_x)
+                curr_pts[3] = curr_pts[2]
+                curr_pts[2] = curr_pts[1]
+                curr_pts[1] = eval_L_curve(new_reg_x)
+            
+            # case 2: right 3 points are better
+                # [a, b, c, d] --> [b, c, b+d-c, d]
+            else:
+                opt_list.append(curr_pts[2])
+                new_reg_x = curr_pts[1].reg_x + curr_pts[3].reg_x - curr_pts[2].reg_x
+                curr_pts[0] = curr_pts[1]
+                curr_pts[1] = curr_pts[2]
+                curr_pts[2] = eval_L_curve(new_reg_x)
+            
+            itr += 1
+            search_range = curr_pts[3].reg_x - curr_pts[0].reg_x
+            if verbose:
+                print('Iteration: %d, deviation of the regularization interval: %.2e'%(itr, search_range))
         
-        S0_stack_f = fftn(S0_stack, axes=(-3,-2,-1))
-        
-        f_real     = np.real(ifftn(S0_stack_f * np.conj(H_eff) / (np.abs(H_eff)**2 + reg_re),axes=(-3,-2,-1)))
+        if verbose:
+            print('Final regularization parameter chosen: %.2e' % opt_list[-1].reg)
+            
+        # Return 3 solutions: the parameter chosen by the L-curve +/- epsilon
+        f_real = []
+        pt_left = eval_L_curve(opt_list[-1].reg_x - epsilon_auto)
+        pt_right = eval_L_curve(opt_list[-1].reg_x + epsilon_auto)
+        f_real.append(ifft_f_real(pt_left.f_real_f))
+        f_real.append(ifft_f_real(opt_list[-1].f_real_f))
+        f_real.append(ifft_f_real(pt_right.f_real_f))
         
-    return f_real
+        return np.array(f_real)
+    
+    else:
+        chosen_pt = eval_L_curve(xp.log10(reg_re))
```

Comment:
When no auto-tuning, we do not need to evaluate the coordinate on L-curve. This can save a bit of computation for non-autotuning situation.